### PR TITLE
Add and update developer guidelines

### LINKED ISSUE #555: Developer guidelines
It'd be useful to have some developer guidelines in our docs, so that we can point contributors to it, and also achieve consensus among ourselves and make our lives easier. This would include at least:
1. Git workflow policies, including merging/rebasing conventions or rules
1. Review workflow, including our unwritten rule that you can't merge your own code (with whatever exceptions we deem appropriate)
1. Style recommendations

### PR DESCRIPTION
**Pull request by berquist**
Part of #555.  Below is a condensed form of the issue contents.  In the interest of time I don't think this PR will cover them all

---

It'd be useful to have some developer guidelines in our docs, so that we can point contributors to it, and also achieve consensus among ourselves and make our lives easier. This would include at least:

- [ ] 1. Git workflow policies, including merging/rebasing conventions or rules
- [ ] 2. Review workflow, including our unwritten rule that you can't merge your own code (with whatever exceptions we deem appropriate)
- [x] 3. Style recommendations
- [x] 4. Procedures for unit test files and code.
- [x] 5. Procedures for regression tests and files in cclib/cclib-data
- [ ] 6. Parsing conventions, such as not predicating parsing on previously parsed attributes. So, parse data multiple times if available, even if we don't use the value for attributes. OTOH, prefer overwriting re-parsed attributes, or at least checking their consistency, via `set_attribute`.

Rebasing formatting changes is making it too difficult to review inside a PR. If we are unhappy with the number of commits in a PR, we should either:
- "Squash and merge"
- Request an interactive rebase to squash those commits just before merging the PR.

- [ ] 7. Describe our approach to parsing files, streams and multiples files and/or directories.

Note for points 4 & 5: when updating unit test files, first copy the newest version of the test logfiles to a new directory and then update them with the actual new contents. That way the history in git shows the changes from the last version by default, which makes things sometimes easier to debug. This was done recently when updating ORCA in https://github.com/cclib/cclib/pull/575 and works nicely.

- [ ] 8. Describe our philosophy for methods: willing to include anything that relies on attributes and physical constants. Methods should also be pure Python and reasonable light.
- [x] From https://github.com/cclib/cclib/pull/568#discussion_r212835180,
> let's have a convention to use only one tool if we want to auto-format. We have enough pep8 violations that it would make sense to format in bulk once if we want to go in that direction.
- [ ] 9. From https://github.com/cclib/cclib.github.io/issues/21, I think we need to outline in the dev section what we expect from pull requests and point people to that when they send very large changes, etc.
- [ ] 10. As mentioned in [Add guide for creating a new parser](https://github.com/cclib/cclib.github.io/issues/15), provide some guidance on how to go about adding new parsers. (Originally from https://github.com/cclib/cclib/pull/257)
- [ ] 11. Explain the mechanics of making new releases.
- [ ] 12. Explain what the goals and non-goals are, in particular that we're not concerned with correctness of data. The goal is to consistently parse data, and not vouch for it.
- [x] 13. Add expectations about using the logger as per [Use of logging utilities](https://github.com/cclib/cclib/issues/237).
- [x] 14. Stick to NumPy style docstrings, or some other format we want, as mentioned in [Add value checking of data, ad add check for negative etenergies](https://github.com/cclib/cclib/pull/694).
- [ ] 15. Preferred function usage:
  - [ ] When skipping lines, prefer using `skip_lines` and `skip_line` with a short message about line contents over repeated calls to `next(inputfile)`.
  - [ ] (see 6) Use `set_attribute("myattr", thing)` and `append_attribute("myattr", thing)` instead of `self.myattr = thing` and `self.myattr.append(thing)` so we can hide convenience things such as checking if we're overwriting an attribute or creating the `myattr` list if it doesn't already exist.
- [x] 16. As part of style (3) and PR size/expectations (9), unrelated changes should in separate PRs, including non-functional changes. Style changes aren't discouraged but they need a separate discussion. Smaller PRs make it easier to do this and for them to be reviewed and merged in general.
- [ ] 17. When updating a PR branch with the latest changes from master or main, prefer to rebase rather than merge.
- [ ] 18. When receiving PR reviews, rebasing makes conversations hard to follow since they are no longer properly associated with their parent commit (this is a GitHub, not Git, problem). Try to hold off on rebasing against the main branch as long as possible.
- [ ] 19. When adding test data for nested programs, where the desired output appears inside the output of another, the target program is the outer folder and the calling program is the inner folder. See [NBO data files for subset of molecules](https://github.com/cclib/cclib/pull/1230) for an example, where the layout is `data/NBO/basicNBO7.0/basicORCA5.0/` for NBO run by ORCA.
- [ ] 20. How to perform development (edit code, run tests) using Dev Containers (see [initial devcontainer.json](https://github.com/cclib/cclib/pull/1212), [Create and test new attributes for the NBO parser](https://github.com/cclib/cclib/pull/1251#issuecomment-1685370562)).

### CODE REVIEW

**Comment by berquist** [comment]
I'll do more after the release is done.

File: doc/sphinx/development.rst [hunk 1]
```diff
@@ -114,10 +156,45 @@ The job types currently included as unit tests:
...
 * static polarizabilities for tryptophan (RHF/STO-3G)
 
+In addition to the above unit tests for data, there are also unit tests for each bridge, calculation method, IO format, and helper utilities, all located inside the `tests <https://github.com/cclib/cclib/tree/master/test>`_ directory, with each category receiving its own subdirectory.
+
 .. _`table of attribute coverage`: data_dev.html#details-of-current-implementation
 
+Adding a new attribute
+^^^^^^^^^^^^^^^^^^^^^^
+
+Definitions of attributes (``mocoefs``, ``natom``, etc.) are located inside the `ccdata <https://github.com/cclib/cclib/blob/0aff0e0d4791f88483c90a63a62e2768794588e9/cclib/parser/data.py#L21>`_ class. Use existing attributes for guidance.
```
Comment (shivupa) [inline doc/sphinx/development.rst:166]: `mocoeffs` typo

**Review by shivupa — APPROVED** [review]
